### PR TITLE
Switched everything to use cryptography instead of pyCrypto

### PR DESCRIPTION
**Pull request by alex**
Motivation:

* Adds PyPy support
* Performance improvement
* OpenSSL and friends are better audited than PyCrypto
* Easier windows install flow (Cryptography provides statically linked wheels
  on Windows)

This PR is basically complete on the code side, of course it can always use
more review :-)

Tests all pass locally (tested with PyPy!)

Still needs some docs work, and to figure out how to do this with the version
numbers so people's stuff doesn't suddenly get broken.

### CODE REVIEW

**Comment by Julian** [comment]
Any chance there's been an update here?

**Comment by drewfisher314** [comment]
I would love to see an update here as well.  Last I heard was the proposed patch to switch to cryptography.io hit a snag in pyasn1, but I might be confused.

**Comment by alex** [comment]
That is correct -- the next release of cryptography will obviate that
problem though. It's due out in less than a week, so I'll update this then.

On Tue, Mar 3, 2015 at 7:14 AM, drewfisher314 <notifications@github.com>
wrote:

> I would love to see an update here as well. Last I heard was the proposed
> patch to switch to cryptography.io hit a snag in pyasn1, but I might be
> confused.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/paramiko/paramiko/pull/394#issuecomment-76963736>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by frewsxcv** [comment]
Worth mentioning that a new version of Cryptography was released a few days ago: https://cryptography.io/en/latest/changelog/#id1

**Comment by alex** [comment]
Yup, I believe it has everything that's needed for this. I'll make time for
it before the end of the week.

On Wed, Mar 11, 2015 at 2:50 PM, Corey Farwell <notifications@github.com>
wrote:

> Worth mentioning that a new version of Cryptography was released a few
> days ago: https://cryptography.io/en/latest/changelog/#id1
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/paramiko/paramiko/pull/394#issuecomment-78344396>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by alex** [comment]
Ok, should be good now!

File: paramiko/ecdsakey.py [hunk 1]
```diff
@@ -20,19 +20,27 @@
...
+from cryptography.exceptions import InvalidSignature
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import ec
+from cryptography.hazmat.primitives.asymmetric.utils import (
+    decode_rfc6979_signature, encode_rfc6979_signature
+)
 
-from paramiko.common import four_byte, one_byte
+from paramiko.common import four_byte, one_byte, zero_byte
```
Comment (reaperhulk) [inline paramiko/ecdsakey.py]: It doesn't look like you're using zero_byte (just b'\x00') in this file.

Comment (alex) [inline paramiko/ecdsakey.py]: Fixed.

**Comment by reaperhulk** [comment]
I've reviewed this and it looks correct. Since this is going to be a 2.0 change would it make sense to consider making more significant changes though? 

For example, right now this patch stores the keys as properties on the key instance and then deserializes them into a key object for every signature call. If the cryptography key object was stored instead there would be less code (to serialize/deserialize) as well as being faster.

Similarly, the RSA/DSA keys use a bespoke BER encoder to serialize keys. This could be potentially replaced with the same code that is being proposed in this PR for ECDSA keys (using cryptography's interfaces).

**Comment by alex** [comment]
I'm torn, and would appreciate input here @bitprophet. The BER ->
cryptography change woudl be straightforward, but it would also make the
diff larger.

On Sun, Mar 15, 2015 at 2:18 PM, Paul Kehrer <notifications@github.com>
wrote:

> I've reviewed this and it looks correct. Since this is going to be a 2.0
> change would it make sense to consider making more significant changes
> though?
>
> For example, right now this patch stores the keys as properties on the key
> instance and then deserializes them into a key object for every signature
> call. If the cryptography key object was stored instead there would be less
> code (to serialize/deserialize) as well as being faster.
>
> Similarly, the RSA/DSA keys use a bespoke BER encoder to serialize keys.
> This could be potentially replaced with the same code that is being
> proposed in this PR for ECDSA keys (using cryptography's interfaces).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/paramiko/paramiko/pull/394#issuecomment-81186758>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by bitprophet** [comment]
I'm +1 on larger diffs as long as most of the largeness is red ;) which if I understand @reaperhulk correctly, would be the case?

Re: release-y things, there's enough other junk on 2.0's plate & it's been in the works long enough, that I am also +1 on small-to-medium additional changes that improve clarity/code size in a backwards incompatible way.

**Comment by alex** [comment]
@bitprophet ok, I took a swing at further reductions in the key modules, let me know what you think (there's still a few small bugs to fix up)

@reaperhulk review please :-)

**Comment by alex** [comment]
Ok, tests are green now.

**Comment by alex** [comment]
@bitprophet re-pining to see where we are with this :-)

**Comment by drewfisher314** [comment]
An update on this would be much appreciated.  Thanks!

File: sites/www/installing.rst [hunk 2]
```diff
@@ -50,61 +48,31 @@ If you're unsure which version to install, we have suggestions:
...
-
-    C:\> pypm install paramiko
-    The following packages will be installed into "%APPDATA%\Python" (2.7):
-     paramiko-1.7.8 pycrypto-2.4
-    Get: [pypm-free.activestate.com] paramiko 1.7.8
-    Get: [pypm-free.activestate.com] pycrypto 2.4
-    Installing paramiko-1.7.8
-    Installing pycrypto-2.4
-    C:\>
+repository or RedHat RPM,, you will also need the ability to build Python
```
Comment (loic) [inline sites/www/installing.rst]: Double comma.

**Comment by bitprophet** [comment]
@alex - hey! Thanks for continuing to poke at this. I've been focused entirely on Invoke and Fabric 2 on the run-up to PyCon, but Paramiko 2.0 is tangentially related to that (e.g. Fabric 2.0.0 final release will quite possibly require Paramiko 2.0+ depending on if I sneak real backwards incompat API changes in).

I wouldn't be surprised if we get this at least merged, if not released, during sprints. Can't promise, but :)

**Comment by alex** [comment]
Awesome! I'll be around for part Monday at the sprints, and semi-available online if it needs any more work then.

**Comment by comay** [comment]
Alex, is there a trick to executing the test cases with this pull request? Along with running into #317, I also see errors concerning ecdsa:

ERROR: test_10_load_ecdsa (tests.test_pkey.KeyTest)
ERROR: test_11_load_ecdsa_password (tests.test_pkey.KeyTest)
ERROR: test_12_compare_ecdsa (tests.test_pkey.KeyTest)
ERROR: test_13_sign_ecdsa (tests.test_pkey.KeyTest)
ERROR: test_2_5_client_ecdsa (test_client.SSHClientTest)
ERROR: test_3_multiple_key_files (test_client.SSHClientTest)

**Comment by alex** [comment]
Nope, tests should be run the same way as usual. It looks like you're on RHEL though, some versions of RHEL have removed ECDSA support from OpenSSL, you might try upgrading your RHEL.

**Comment by comay** [comment]
Actually, testing the pull request on Solaris but thanks for pointing me in the right direction - this particular version of OpenSSL doesn't support ECDSA.

**Comment by alex** [comment]
@bitprophet A heads up, we're going to be dropping Python 3.2 support from the next release of cryptography because the usage is *so* low. Can you confirm if this will be a problem for Paramiko?

**Comment by drewfisher314** [comment]
Any word on a new paramiko release w/ cryptography?  Obviously a few of us are very anxiously awaiting it.

**Comment by bitprophet** [comment]
@alex 3.2 support is ambiguous right now, our README says 3.3+, setuptools classifiers say 3.2, install doc says "probably works on 3.2 but the guarantee is not a strong one", our Travis setup runs on 3.2 (partly...it skips the doc build because Sphinx also doesn't do 3.2).

tl;dr I'm probably fine with it, I'd certainly rather have crypto.io and no 3.2 than 3.2 and be stuck with pycrypto :D 

@drewfisher314 If there's news it'll be in here. This is still the top of my stack for Paramiko, I've been striving for a milestone in a related project in the meantime; once I get there this is the next focus.

**Comment by alex** [comment]
FWIW, @dstufft has some numbers that show that basically no one uses 3.2
with paramiko :D

On Tue, May 5, 2015 at 1:34 PM, Jeff Forcier <notifications@github.com>
wrote:

> @alex <https://github.com/alex> 3.2 support is ambiguous right now, our
> README says 3.3+, setuptools classifiers say 3.2, install doc says
> "probably works on 3.2 but the guarantee is not a strong one", our Travis
> setup runs on 3.2 (partly...it skips the doc build because Sphinx also
> doesn't do 3.2).
>
> tl;dr I'm probably fine with it, I'd certainly rather have crypto.io and
> no 3.2 than 3.2 and be stuck with pycrypto :D
>
> @drewfisher314 <https://github.com/drewfisher314> If there's news it'll
> be in here. This is still the top of my stack for Paramiko, I've been
> striving for a milestone in a related project in the meantime; once I get
> there this is the next focus.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/paramiko/paramiko/pull/394#issuecomment-99153324>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by bitprophet** [comment]
I was tempted to bug him for that but I'm glad you already did, thanks :D

**Comment by alex** [comment]
@bitprophet I updated this PR with the python 3.2 stuff discussed several moons ago.

**Comment by drewfisher314** [comment]
What's the current status of this PR?  It seems to have gotten forgotten :(

**Comment by bitprophet** [comment]
@drewfisher314 it's called a day job :( this is still totally in the top few items on my list, it's just big enough (alongside the other top items like uggggh key handling) that I need to dedicate sprint time to it. Planning to spend early September in "paramiko mode".

**Comment by drewfisher314** [comment]
I understand completely (I also have one of those pesky day job things).  Thanks for the update!

**Comment by alex** [comment]
Resolved a couple of merge conflicts that had snuck in.

**Comment by techtonik** [comment]
All checks have failed.

**Comment by bitprophet** [comment]
At least one of those is probably due to me nuking 3.2 stuff from Travis because I got fed up with it breaking things. (As a corollary I am definitely now cool with the overall removal of 3.2 support in this PR and thus Paramiko 2.0...just for the record, since I was on the fence earlier.) The others are likely to be due to the recently merged SHA-2 support.

Speaking of which, those merges mean Paramiko 1.16 is almost done, then 2.0 is the next thing up, finally.

**Comment by alex** [comment]
@bitprophet I'm unable to reproduce any of these failures locally, and they seem to be py3k specific. Ever seen anything like this?

**Comment by techtonik** [comment]
@alex turn more verbose log / test trace on Travis? Might be a limitation of Travis Linux container.

**Comment by bitprophet** [comment]
Sorry I was referring to the merge conflicts, not the test failures - didn't see those at first.

Peeked at Travis and both of those issues (10-minute timeout & `test_L_handshake_timeout` failing) are ones I see on Travis frustratingly often for our main branches (tho most often under Python 3.2 which was on reason I nixed it from the old branches myself).

I'm also unable to get either of those to happen for me on my development workstation.

Clearly it's getting to the point where I need to take extra time to reproduce (hopefully a Linux VM + looping over the tests a bunch will suffice) though solving maybe another matter.

**That said**, these issues are not specific to this branch so I am not worried about the efficacy of the PR's changes, especially if the tests pass for me when I get around to merging.

Speaking of, Paramiko 1.16 is now out so as stated the next chunk of work for me on Paramiko is 2.0. I'll be hacking on that and finishing the alpha of Fabric 2.0 around the same time.

**Comment by alex** [comment]
Awesome, let me know if there's ever anything you need from me on finishing
this up.

On Thu, Nov 5, 2015 at 7:24 PM, Jeff Forcier <notifications@github.com>
wrote:

> Sorry I was referring to the merge conflicts, not the test failures -
> didn't see those at first.
>
> Peeked at Travis and both of those issues (10-minute timeout &
> test_L_handshake_timeout failing) are ones I see on Travis frustratingly
> often for our main branches (tho most often under Python 3.2 which was on
> reason I nixed it from the old branches myself).
>
> I'm also unable to get either of those to happen for me on my development
> workstation.
>
> Clearly it's getting to the point where I need to take extra time to
> reproduce (hopefully a Linux VM + looping over the tests a bunch will
> suffice) though solving maybe another matter.
>
> *That said*, these issues are not specific to this branch so I am not
> worried about the efficacy of the PR's changes, especially if the tests
> pass for me when I get around to merging.
>
> Speaking of, Paramiko 1.16 is now out so as stated the next chunk of work
> for me on Paramiko is 2.0. I'll be hacking on that and finishing the alpha
> of Fabric 2.0 around the same time.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/paramiko/paramiko/pull/394#issuecomment-154242150>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by bitprophet** [comment]
Absolutely, thanks as usual @alex and apologies this has taken so long >_<

**Comment by alex** [comment]
merged master in because of merge conflicts. the test failure is due to #612

**Comment by techtonik** [comment]
So, ping Travis again?

**Comment by bitprophet** [comment]
@techtonik #612 is still unresolved but its tl;dr is "Travis triggers race conditions often under Python 3" so just tickling it isn't usually enough to get a good sense of whether a PR works. Will finalize #612 before this gets merged.

**Comment by johnthagen** [comment]
Just curious on the status of this PR. As someone who supports Windows, very excited to have `cryptography`'s official Windows wheels on PyPI.  Appreciate the work put into this.

**Comment by alex** [comment]
Resolved a merge conflict that had crept in.

**Comment by alex** [comment]
(test failure unrelated)

**Comment by techtonik** [comment]
@alex restart [test](https://travis-ci.org/paramiko/paramiko/builds/112339521)?

**Comment by johnthagen** [comment]
@alex @bitprophet Just FYI, not sure if this affects `paramiko` directly, but `PyCrypto` has a known exploitable buffer overflow https://github.com/dlitz/pycrypto/issues/173.  Could increase the priority of porting to `cryptography`.

**Comment by bitprophet** [comment]
@johnthagen Eek :( good to know. I was planning to expedite this anyways for other reasons (not least "I feel like a tremendous ass for letting it moulder so long"), but that's another excellent one.

----

So I finally got my ass in gear on #612 and *think* I have it licked so it'll stop making this PR (among others) harder to test.

Made my own integration branch for this ticket, and after a `pip install cryptography>=0.8` I note the following things:

* Cryptography's added a handful more sub-dependencies than I remember: on top of `pyasn1` there's now `pycparser`, `cffi`, and `idna`. Presuming these are all still pure-Python & relatively stable, and given the payoff of merging this, I don't think I'm going to complain much. Just want it on the record.
* I'm getting 3 failures/errors all from the same basic code path re: loading an RSA key (posted below).
    * Given that my setup (OS X Yosemite) isn't uncommon, this does make me worry a bit, but I'm confident it's solvable and probably something dumb like cruft in my environment somewhere.
    * If someone else sees this before I get around to debugging it and has suggestions/a fix, I am all ears. I plan to tackle this once I dig through the rest of https://github.com/paramiko/paramiko/milestones/1.16.1
* Travis is **not** seeing those errors - it's happy: https://travis-ci.org/paramiko/paramiko/builds/125184508 
----

```
======================================================================
ERROR: test_client_can_be_used_as_context_manager (test_client.SSHClientTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_client.py", line 318, in test_client_can_be_used_as_context_manager
    self.tc.connect(self.addr, self.port, username='slowdive', password='pygmalion')
  File "/Users/jforcier/Code/oss/paramiko/paramiko/client.py", line 367, in connect
    look_for_keys, gss_auth, gss_kex, gss_deleg_creds, gss_host)
  File "/Users/jforcier/Code/oss/paramiko/paramiko/client.py", line 558, in _auth
    key = pkey_class.from_private_key_file(filename, password)
  File "/Users/jforcier/Code/oss/paramiko/paramiko/pkey.py", line 196, in from_private_key_file
    key = cls(filename=filename, password=password)
  File "/Users/jforcier/Code/oss/paramiko/paramiko/rsakey.py", line 47, in __init__
    self._from_private_key_file(filename, password)
  File "/Users/jforcier/Code/oss/paramiko/paramiko/rsakey.py", line 166, in _from_private_key_file
    self._decode_key(data)
  File "/Users/jforcier/Code/oss/paramiko/paramiko/rsakey.py", line 174, in _decode_key
    data, password=None, backend=default_backend()
  File "/Users/jforcier/.virtualenvs/paramiko3/lib/python3.5/site-packages/cryptography/hazmat/primitives/serialization.py", line 28, in load_der_private_key
    return backend.load_der_private_key(data, password)
  File "/Users/jforcier/.virtualenvs/paramiko3/lib/python3.5/site-packages/cryptography/hazmat/backends/multibackend.py", line 307, in load_der_private_key
    return b.load_der_private_key(data, password)
  File "/Users/jforcier/.virtualenvs/paramiko3/lib/python3.5/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 1122, in load_der_private_key
    password,
  File "/Users/jforcier/.virtualenvs/paramiko3/lib/python3.5/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 1265, in _load_key
    self._handle_key_loading_error()
  File "/Users/jforcier/.virtualenvs/paramiko3/lib/python3.5/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 1337, in _handle_key_loading_error
    raise ValueError("Could not unserialize key data.")
ValueError: Could not unserialize key data.
```

**Comment by alex** [comment]
- The `idna` dep is newish, `cffi` and `pycparser` we've had the whole time.
- Not sure what to make of this traceback; are you able to reproduce this issue easily locally?

**Comment by alex** [comment]
@bitprophet Ok, so that failure is probably related to your `~/.ssh/id_rsa` (which I do not have locally, which is probably why I can't reproduce).

Not sure how to proceed in debugging without seeing the contents of your `id_rsa`, which for SUPER obvious reasons you wouldn't want to give me.

**Comment by bitprophet** [comment]
Yea, that would deffo fall under "my local environment" huh. I peeked and the fact that it's being loaded at all is just due to default parameter val for `Client.connect(look_for_keys)`. Why this wasn't causing test failures before the switch is a good question, I didn't think my RSA key was that off-spec (been using it for ages, including against relatively modern distros). I welcome ideas there and will poke otherwise.

Also gonna explicitly set `look_for_keys=False` on these tests because this is poor behavior for a test suite IMHO.

**Comment by alex** [comment]
Unfortunately I can't think of a good way for me to debug without you
sharing the key. Can you try generating a fresh key, putting it there, and
if that triggers the error sharing _that_ key?

On Sat, Apr 23, 2016 at 12:39 PM, Jeff Forcier <notifications@github.com>
wrote:

> Yea, that would deffo fall under "my local environment" huh. I peeked and
> the fact that it's being loaded at all is just due to default parameter val
> for Client.connect(look_for_keys). Why this wasn't causing test failures
> before the switch is a good question, I didn't think my RSA key was that
> off-spec (been using it for ages, including against relatively modern
> distros). I welcome ideas there and will poke otherwise.
>
> Also gonna explicitly set look_for_keys=False on these tests because this
> is poor behavior for a test suite IMHO.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/paramiko/paramiko/pull/394#issuecomment-213777284>
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: D1B3 ADC0 E023 8CA6

**Comment by bitprophet** [comment]
Yea, I'll do that, and also going to try comparing what I can between it & some of the keys stored in the test suite.

Amusingly I was going to regenerate new keys for myself soon anyways, but I figure if a key I presumably generated in a vanilla fashion a while back is "broken", other users might encounter this too, so still want to figure out WTF.

**Comment by bitprophet** [comment]
Digging more, suspect the reason this fails on this branch and not master isn't the key, it's the type of exception raised; most of the internals of Paramiko have `try / except SSHException` sprinkled around, and failure to load a passphrase-protected key raises `SSHException` subclasses; so such exceptions get caught, stored, and then only if all attempts to auth fail, are any of those bubbled up. (This is very much re: #387).

In the test suite, it's almost all password auth, so what happens is keys are tried & fail & are muted, then password auth is tried, succeeds, and the earlier errors are discarded.

In this PR's code, the attempt to load the personal/protected key is raising `ValueError`, which Paramiko isn't tuned to expect, so it bubbles up immediately, not allowing the "trickle down" to password auth to occur.

I'll look deeper now to see what the best solution probably is. I loathe `SSHException` but was hoping to save tweaking it for later, really don't want it to be a blocker on this. Hoping we can extend a low-level `except` to also skip/store `ValueError` & pray that doesn't make it cast too wide a net.

**Comment by alex** [comment]
Ahhhh, and I bet there's no tests for "failure to parse raises an SSHError", it sounds like the key loading should just be written in a `try` / `except ValueError: raise SSHError`?

**Comment by bitprophet** [comment]
Yea pretty much. We'll see!

**Comment by bitprophet** [comment]
Also, you're so fast on responding to updates that I need to curb my "ninja edit" habit. :D :D plz do see the parenthetical above about release stuff.

**Comment by bitprophet** [comment]
Also, shoutout to how the `ValueError`-raising statement is on line `1337` of the openssl backend. :+1:

**Comment by alex** [comment]
@bitprophet should be fully resolved now!

**Comment by bitprophet** [comment]
Looks ok, testing locally :) We ideally want a more useful message in the exception than "Unable to deserialize" (something saying "I tried deserializing this key *with this password* and it failed") but I think that can be considered part of #387.

Anyway...

* Nabbed latest, confirm test suite still happy (because of my earlier changes)
* Confirmed my temp reproduction code triggering the issue is now happy (insofar as it raises the same SSHException based "couldn't make any auth work" exception chain as master)
    * Temp-reverted my earlier fixes to the test suite re: `look_for_keys`, confirmed test suite *still* happy.
* Created a new, real test because manual testing is dumb & it makes the test suite that much closer to a description of expected behavior
    * Annoyingly, this ValueError business doesn't happen with the test suite's password-protected *DSA* key, but does happen with the *RSA* key. Was not expecting this, started out using the DSA key arbitrarily, and that wasted a lot of my time :cry:
    * Tested against this PR branch prior to alex's recent fixes; confirmed fail
    * Tested against latest HEAD of this branch; confirmed pass
    * Tested against master; confirmed pass

**Comment by alex** [comment]
@bitprophet Awesome. What's the next step here?

**Comment by Julian** [comment]
Champagne. Is it champagne time? I'll buy?

**Comment by bitprophet** [comment]
@alex @Julian I deffo want to get this out ASAP, what remains as far as I can tell:

* I gotta make sure I personally understand the installation & upgrade ramifications, both generally (I had the mistaken assumption this was all pure-Python and not C extensions. I blame me being stupid.) and re: the handful of newer dependencies.
    * I expect this to just be me testing installation a few times in a few envs, & maybe tweaking the install docs a bit.
* Need to decide whether this lives in 1.17 or 2.0 as previously planned. Still leaning 2.0, but in a sense where the rest of the current 2.0 milestone gets bumped to 3.0 (with 3.0 being slated for "still pretty soon").
    * tl;dr I want this to get out in the wild without having to wait for the technically-unrelated other stuff like #387, which could take more time to hammer out
    * but there's still that ingrained reluctance to put out "too many" major revisions "too soon". I think this is mostly bunk though, esp since a 2.0 with just this PR merge is fully API compatible with 1.0.
        * But using major versioning to signal "a big potentially disruptive change" still seems like a better idea than "oh yea this is just another 1.N feature release".
* Want to get 1.16.1 out first to relieve bugfix pressure on the 1.x line. Shouldn't take more than a day or two (& if it does I will just punt stuff).
    * Will probably merge #394 to master before then, just to make it easier for bleeding edge folks to try it out.

**Comment by reaperhulk** [comment]
@bitprophet Yeah there are definitely some nuances since cryptography links against OpenSSL (and CommonCrypto on OS X). A few things to keep in mind:

* We distribute binary wheels for OS X and Windows, so users there don't even need a compiler installed. However, wheel started using a 2.7 SOABI tag recently so users need to have pip 8.1+ to properly download the wheel on OS X. An easy fix if encountered, but something to be aware of (https://cryptography.io/en/latest/faq/#compiling-cryptography-on-os-x-produces-a-fatal-error-openssl-aes-h-file-not-found-error)
* On Linux/FreeBSD/etc users will need a C compiler as well as OpenSSL development headers if they want to compile it themselves.

Cryptography is currently planning to drop OpenSSL 0.9.8 support in its next release as well. This should only affect extremely old distributions such as RHEL/CentOS 5, but we don't want to catch anyone by surprise. For the 1.4 release (probably released next month) there will be a [workaround](https://cryptography.io/en/latest/faq/#importing-cryptography-causes-a-runtimeerror-about-openssl-0-9-8), but if you feel strongly about continuing 0.9.8 support let us know :)

**Comment by bitprophet** [comment]
Notes to myself...

### Install testing

* Installing `cryptography` in a vanilla Python 2.7 (OS X Yosemite bundled version; I have the usual Python build headers & openssl headers installed too) virtualenv using pip 7.1.2 results in the following:

```
» pip freeze
cffi==1.6.0
cryptography==1.3.1
enum34==1.1.3
idna==2.1
ipaddress==1.0.16
pyasn1==0.1.9
pycparser==2.14
six==1.10.0
wheel==0.24.0
```

* `cryptography` and `cffi` download as `tgz` and then run `bdist_wheel`; the rest download as `whl` directly.
* After blowing away my virtualenv, recreating it, and upgrade pip to latest 8.x (EDIT: I noticed this myself from crypto's docs, but @reaperhulk confirms above too, thx!), confirm that both `cryptography` and `cffi` download & install wheels directly instead.
* On a Debian 8 vm which previously had python-dev installed, also with pip 7.1.2, as expected had to install `libffi-dev` and `libssl-dev` for those two wheels to build successfully.
* Further, confirmed that a workable up-to-date dev virtualenv for paramiko master passes tests, then checked out this branch, `pip install -e .` to receive the upgraded dependencies, then confirmed tests continue to pass.
    * And nuking pycrypto from the virtualenv has no observable result, as expected.

### Stuff to tweak in paramiko

(I can handle these...)

* [ ] Because the install docs are part of the www and not docs site, I'll need to shuffle things around so the 1.x version of the doc is still available in some form.
    * Could be an argument for moving the install doc to the (versioned) docs site...
    * Otherwise, simply moving 1.x specific stuff to another page or section, should suffice. I don't expect *installation* to change much after this, even in 3.0 or 4.0 or whatnot - those will be API or behavioral changes.
* [ ] Previously, `pyasn1` is an optional dependency, for the GSSAPI stuff. That language needs to be removed now that `pyasn1` is a hard requirement as a sub-dependency of crypto.io.
    * [ ] This branch lists `pyasn1` as a hard dep, but I think it *may* be easier for all concerned to simply defer to the crypto.io docs (with a warning that it includes "a handful of C extensions and pure-python modules" or w/e) instead of trying to mirror the full list of sub-dependencies, having them get out of date, etc.
    * [ ] Another good argument for deferring is that the language in this PR's version of the install doc, about needing to build from source, lacks the updated info (again per crypto's docs and @reaperhulk) about pip 8 enabling static wheels on OS X. So if we don't defer, this needs an update too.
* [ ] Consider adding an "upgrading from 1 to 2" section to the install docs; this would also make it a good place to note that `ecdsa` is no longer required (otherwise that should at least go in the changelog?)
* [ ] Not 100% sure how to handle the PyPM section given PyPM may not upgrade their side of things immediately (?) but leaning "leave it as-is, direct any complaints to PyPM if it comes up"

**Comment by bitprophet** [comment]
Poking the docs now...almost forgot one thing though. @reaperhulk / @alex I'm wondering if there's a way to avoid the potential 'shock' of the 0.9.8 version drop somehow (but no I don't think its support should be kept, offhand).

* If I get Paramiko 2.0 out in the next few days, and leave its setup data saying `cryptography>=0.8`, this means users would be capable of installing it + crypto 1.3 + openssl 0.9.8, then getting screwed the next time they reinstall their env or `pip install -U`.
* I could switch to `cryptography >=0.8,<1.4`, which would prevent that scenario initially.
    * I imagine following this up by changing that restriction, with a warning in the changelog, in e.g. 2.1 or 2.2. Lets us move away from openssl 0.9.8 but with somewhat less surprise.
* Alternately: stick with `>=0.8`, but put a warning in my own docs mirroring Crypto's, noting "heyyy openssl 0.9.8 sucks don't use it or you'll be sorryyyy".
    * Asses covered, less changes required down the road. Still potentially surprising for folks who don't read docs. Meh?
* Alternately again: wait until Crypto 1.4 is out before releasing Paramiko 2.0.
    * This *might* also let me have time to slap more backwards incompat tickets into Paramiko master
    * Downside: means folks waiting for this poor ticket are waiting even longer; slight danger of me being pulled into other things in the interim.
* Worst-case scenario: like half my users hear about these plans re: OpenSSL 0.9.8, throw a pity party about it because they're still on CentOS 5 or whatever, and I ponder leaning into your workaround or requesting longer-term 0.9.8 support.
    * I don't see this being that likely; I *do* have a surprisingly large number of folks stuck on Python 2.6, but that by definition excludes users on stock CentOS 5 (which is Python 2.4). If someone is on CentOS 5 and backported Python 2.6, surely they can backport OpenSSL.

**Comment by alex** [comment]
FWIW, by the numbers (NB: we only have OpenSSL version data for folks on a very recent pip): In the past 2 weeks, `0.3%` of installs of `paramiko` used OpenSSL 0.9.8 (161,004 download sample).

**Comment by reaperhulk** [comment]
I'm in favor of option 3 myself. 0.9.8 is over 8 years old and its mere existence is holding back the security of the internet :angry:

**Comment by bitprophet** [comment]
Works for me, thanks both!

Mostly done with the docs treatment now, will push my integration branch when that's complete, though that's more for reference, I don't care hard whether you merge it, I can just as easily merge it + this into master when necessary.

**Comment by bitprophet** [comment]
Neglected to tag this on another commit re: docs...here's a SHA-pinned comparison, cc @alex @reaperhulk: https://github.com/alex/paramiko/compare/alex:d1e43c1...paramiko:0ae4d44

**Comment by reaperhulk** [comment]
Mostly looks good. Only comment I have is that "plus development headers for Python, OpenSSL and CFFI" should probably read "plus development headers for Python, OpenSSL and libffi", as libffi is cffi's C dependency.

**Comment by bitprophet** [comment]
Updated, thanks!

**Comment by bitprophet** [comment]
Posted http://bitprophet.org/blog/2016/04/23/paramiko-2.0-is-coming/ so this doesn't catch everyone 100% flat-footed when it comes out :3

I'm off to bang on 1.16.1; some of its tickets may get bumped to 1.17.0; once those are both squared away I am probably going to declare master "2.0" and merge this.

**Comment by alex** [comment]
\o/

Thanks! Feel free to CC me on any issues should there be follow up necessary.

**Comment by bitprophet** [comment]
Did initial prep work for release:

* Branched old master as `1.17` so that can be released (it's tiny, 1.16.1 will be much bigger changelog-wise)
* Updated my integration branch (which has some of its own changes, as above, mostly docs) with alex's branch and master (been working on bugfixes a lot today). Looks ok
* Merged that to master and pushed, which is why this is merged now :3

I plan to release 1.16.1, 1.17.0 and 2.0.0 tomorrow if all goes well. Thanks a billion, Alex, you have the patience of a saint.

**Comment by bitprophet** [comment]
> Feel free to CC me on any issues should there be follow up necessary.

You know I will 😆

**Comment by drewfisher314** [comment]
Outstanding!

**Comment by techtonik** [comment]
:+1:

**Comment by Julian** [comment]
@alex @bitprophet and others, all y'all are great. Thanks!

**Comment by bitprophet** [comment]
Quick note, still planning to get this out ASAP, been banging on an update to my changelog library so it can actually support 1.0 -> 2.0 transitions gracefully. *Mostly* done with that now. Computers.

**Comment by alex** [comment]
@bitprophet Sounds good, let me know if you need anything.

**Pull request merged**